adr/ADR-32.md

```diff
@@ -0,0 +1,179 @@
+\* | `js-consumer-info` | `$JS.API.CONSUMER.INFO.*.*` | pub |
+\* | `js-consumer-info(stream)` | `$JS.API.CONSUMER.INFO.{stream}.*` | pub |
+\* | `js-consumer-info(stream, name)` | `$JS.API.CONSUMER.INFO.{stream}.{name}` | pub |
+\* | `js-consumer-info(*, name)` | `$JS.API.CONSUMER.INFO.*.{name}` | pub |
+\* | `js-consumer-next-msg` | `$JS.API.CONSUMER.MSG.NEXT.*.*` | pub |
+\* | `js-consumer-next-msg(stream)` | `$JS.API.CONSUMER.MSG.NEXT.{stream}.*` | pub |
+\* | `js-consumer-next-msg(stream, name)` | `$JS.API.CONSUMER.MSG.NEXT.{stream}.{name}` | pub |
+\* | `js-consumer-next-msg(*, name)` | `$JS.API.CONSUMER.MSG.NEXT.*.{name}` | pub |
+\* | `js-consumer-ack-reply` | `$JS.ACK.*.*.>` | pub |
+\* | `js-consumer-ack-reply(stream)` | `$JS.ACK.{stream}.*.>` | pub |
+\* | `js-consumer-ack-reply(stream, name)` | `$JS.ACK.{stream}.{name}.>` | pub |
+\* | `js-consumer-ack-reply(*, name)` | `$JS.ACK.*.{name}.>` | pub |
+KeyValue | `kv-put` | `$KV.*.>` | pub
+\* | `kv-put(bucket)` | `$KV.{bucket}.>` | pub
+\* | `kv-put(bucket, key)` | `$KV.{bucket}.{key}` | pub
+\* | etc..
+
+*TODO: add remaining subjects, leader, peer, advisories, etc.*
+
+The second logical tier can be permission groups, often designed as *roles*, such as `js-stream-operator`. Parameters, such as stream or consumer names, will be transitively applied to the underlying permissions. The goal of roles is to abstract away a set of permissions for common use cases.
+
+This table shows a few examples.
+
+Name | Permissions | Notes
+---- | ----------- | -----
+`js-stream-operator` | `js-create-stream`, `js-update-stream`, `js-delete-stream` `js-purge-stream`, `js-stream-info`, `js-snapshot-stream`, `js-restore-stream`
+`js-stream-operator(name)` | `js-create-stream(name)`, `js-update-stream(name)`, etc...
+`js-stream-user(stream, subject)` | `publish(subject)`, `js-stream-info(stream)`, `js-stream-get-last-subject-msg(stream, subject)` | The `publish(subject)` is alternative to independently setting a pub-allow.
+
+
+The best way to model this in configuration is TBD, however, one approach could support `allow` and `deny` as top-level keys in the `permissions` map which would be used exclusively with these new logical permissions.
+
+For example:
+
+```
+accounts: {
+  APP: {
+    users: [
+      {
+        user: operator,
+        password: operator,
+        permissions: {
+          allow: [
+            "inbox(operator)",
+            "js-stream-operator(EVENTS)",
+          ]
+        }
+      },
+      {
+        user: greeter,
+        password: greeter,
+        permissions: {
+          allow: [
+            "inbox(greeter)",
+            "sub(services.greeter)",
+            "pub(events.greeter.>)",
+          ],
+          allow_responses: true,
+        }
+      },
+      {
+        user: joe,
+        password: joe,
+        permissions: {
+          allow: [
+            "inbox(joe)",
+            "pub(services.*)",
+          ]
```

Comment:
this syntax is nice, I was thinking of granular permission like below but seems very verbose:

```hcl
    permissions = {
      # Can publish to stream
      publish: { allow = ["a"] } 

      # Granual permissions for JetStream features
      jetstream: {
        streams { EVENTS = { allow: [ stream-info ], deny: [add-stream, delete-stream] } }
        kv { TEST = { allow: [ put, del], deny: [ destroy-kv ] }
        object-store { allow: [...] }
      }
    }
   }
```

Reply:
I like the structure that your example provides which provides namespacing (as I am doing `js-` and `kv-`). A future state I considered was user-defined permissions which may break the structure. For example, if a new key called `roles` within an account is defined, you could declare new mappings off the primitive logical permissions the NATS server provides.

Reply:
It should be possible to subset keys/object names on kv/object store - currently object store doesn't expose data chunks in a way that can be clamped.